File: Rust_Solutions/solution4.rs
use std::env;

// User input
fn reverse_string(s: &str) -> String {
    s.chars().rev().collect()
}
// End of user input

struct Checker;

impl Checker {
    fn check(func: fn(&str) -> String, input: &str, expected: &str) {
        let result = func(input);
        if result == expected {
            println!("Test Case Passed: Result = {}, Expected = {}", result, expected);
        } else {
            println!("Failed Test Case: Result = {}, Expected = {}", result, expected);
        }
    }
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() != 3 {
        eprintln!("Usage: {} <string> <expected>", args[0]);
        std::process::exit(1);
    }

    let input = &args[1];
    let expected = &args[2];

    Checker::check(reverse_string, input, expected);
} 

File: Solutions/C_Solutions/solution4.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

void reverseString(char* str) {
    int n = strlen(str);
    for (int i = 0; i < n / 2; i++) {
        char temp = str[i];
        str[i] = str[n - i - 1];
        str[n - i - 1] = temp;
    }
}

void Checker(void (*func)(char*), char* input, char* expected) {
    func(input);
    if (strcmp(input, expected) == 0) {
        printf("Test Case Passed: Result = %s, Expected = %s\n", input, expected);
    } else {
        printf("Failed Test Case: Result = %s, Expected = %s\n", input, expected);
    }
}

int main(int argc, char *argv[]) {
    if (argc != 3) {
        fprintf(stderr, "Usage: %s <string> <expected>\n", argv[0]);
        return 1;
    }
    char* str = argv[1];
    char* expected = argv[2];
    Checker(reverseString, str, expected);
    return 0;
} 